src/ukify/ukify.py

```diff
@@ -1139,6 +1151,42 @@ def parse_hwid_dir(path: Path) -> bytes:
     return devices_blob + strings_blob
 
 
+def parse_efifw_dir(path: Path) -> bytes:
+    if not path.is_dir():
+        raise ValueError(f'{path} is not a directory or it does not exist.')
+
+    # only one firmware image must be present in the directory
+    # to uniquely identify that firmware with its ID.
+    if len(list(path.glob('*'))) != 1:
+        raise ValueError(f'{path} must contain exactly one firmware image file.')
+
+    payload_blob = b''
+    for fw in path.iterdir():
+        payload_blob += fw.read_bytes()
+
+    payload_len = len(payload_blob)
+    if payload_len == 0:
+        raise ValueError(f'{fw} is a zero byte file!')
+
+    dirname = path.parts[-1]
+    # firmware id is the name of the directory the firmware bundle is in,
+    # terminated by NULL.
+    fwid = b'' + dirname.encode() + b'\0'
+    fwid_len = len(fwid)
+    magic = bytes.fromhex(FWHEADERMAGIC)
+
+    efifw_header_blob = b''
+    efifw_header_blob += struct.pack('<p', magic)
+    efifw_header_blob += struct.pack('<I', EFIFW_HEADER_SIZE)
+    efifw_header_blob += struct.pack('<I', fwid_len)
+    efifw_header_blob += struct.pack('<I', payload_len)
+
+    efifw_blob = b''
+    efifw_blob += efifw_header_blob + fwid + payload_blob
+
+    return efifw_blob
```

Comment:
This loading and processing is done in the wrong place. The option parsing code is **not** supposed to do any work or even access the file system. The verification of paths is deferred to `check_inputs` and later stages.

Reply:
Not sure what you are saying! `check_inputs()` is called before `make_uki()` which then calls `parse_efifw_dir()` . There are some generic validations that happen in `check_inputs()` which is basically whether the directory can be opened or not. Then efifw specific validations should happen. That is the job of `parse_efifw_dir()` . This is similar to `parse_hwid_dir()` .

Reply:
What we maybe able to do is move the validations to a seperate function and call it separately from `make_uki()` and then leave the rest as is. That is, something like:
```
if opts.efifw:
  validate_efifw_opt(fw) for fw in opts.efifw
```
then,
```
efifw_blob = parse_efifw_dir(fw) for fw in opts.efifw
```
something like that.

Reply:
Ah, OK. I misread the code.